File: wehe/src/main/java/mobi/meddle/wehe/ui/replay/ReplayActivity.java
package mobi.meddle.wehe.ui.replay;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.os.Bundle;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;
import java.util.Objects;
import mobi.meddle.wehe.R;
import mobi.meddle.wehe.adapter.ImageReplayRecyclerViewAdapter;
import mobi.meddle.wehe.data.bean.ApplicationBean;

/**
 * Runs the replays.
 * XML layout: activity_replay.xml
 * adapter.ImageReplayRecyclerViewAdapter.java for layout of each replay
 */
public class ReplayActivity extends AppCompatActivity {
    public static final String STATUS = "ReplayActPrefsFile";
    private boolean replayOngoing = false;
    private ArrayList<ApplicationBean> selectedApps = null; //apps to run
    private final ArrayList<ApplicationBean> diffApps = new ArrayList<>(); //apps with differentiation
    private final ArrayList<ApplicationBean> inconclusiveApps = new ArrayList<>();
    private ProgressBar prgBar; //progress bar at bottom of screen when tests are running
    private ImageReplayRecyclerViewAdapter adapter = null; //layout for each replay
    private Context context;
    private TraceRunAsync traceRunner; //runs the tests
    private boolean runPortTests;
    private boolean isIPv6; //true if user's public IP is v6, use to display in results
    private String carrier; //carrier to display in results
    private String serverDisplay; //server to display in the results
    private boolean mlabServerUsed;
    //Tomography tests determine where exactly in the network differentiation occurs. If differentiation
    //is detected in a test, the app will ask users if they want to run a tomography test. These
    //tests run 3 concurrent tests to 3 optimal MLab servers. Based on the times the packets are sent,
    //an algorithm can determine where differentiation occurs. All 3 of the tests count as one Wehe
    //"Test", so 1 historyCount is used for all 3 tests.
    private boolean isTomography = false; //true if tomography test, false if normal test
    private final DialogInterface.OnClickListener doNothing = new DialogInterface.OnClickListener() {
        @Override
        public void onClick(DialogInterface dialog, int which) {
        }
    };
    //this happens if rerun differentiation or rerun inconclusive buttons clicked
    private final DialogInterface.OnClickListener rerunButtons = new DialogInterface.OnClickListener() {
        @Override
        public void onClick(DialogInterface dialog, int which) {
            //change page title
            Objects.requireNonNull(getSupportActionBar()).setTitle(getString(R.string.replay_page_title));
            if (which == DialogInterface.BUTTON_POSITIVE) {
                selectedApps = new ArrayList<>(diffApps);
            } else if (which == DialogInterface.BUTTON_NEGATIVE) {
                selectedApps = new ArrayList<>(inconclusiveApps);
            }
            //rearrange layout to hide rerun button
            RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)
                    findViewById(R.id.appsRecyclerView).getLayoutParams();
            params.addRule(RelativeLayout.ABOVE, R.id.prgBarLayout);
            findViewById(R.id.rerunButton).setVisibility(View.GONE);
            findViewById(R.id.localizeDiffButton).setVisibility(View.GONE);
            adapter.setTomography(false);
            isTomography = false;
            for (ApplicationBean app : selectedApps) {
                app.setTomography(false);
                app.setArcepNeedsAlerting(false);
                app.setAlertFCC(false);
                app.setStatus(getString(R.string.pending));
            }
            inconclusiveApps.clear();
            diffApps.clear();
            traceRunner = new TraceRunAsync(ReplayActivity.this);
            traceRunner.execute("");
        }
    };
    //rerun dialogue
    private final View.OnClickListener rerunListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            //dialogue box to rerun tests
            AlertDialog.Builder alertDialog = new AlertDialog.Builder(ReplayActivity.this,
                    AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
                    .setTitle(R.string.rerun_test_title)
                    .setMessage(R.string.rerun_test_descr);
            //rerun tests with differentiation
            if (diffApps.size() != 0) {
                alertDialog.setPositiveButton(R.string.rerun_diff_opt, rerunButtons);
            }
            //rerun only the inconclusive tests; doesn't appear if no tests inconclusive
            if (inconclusiveApps.size() != 0) {
                alertDialog.setNegativeButton(R.string.rerun_incon_opt, rerunButtons);
            }
            alertDialog.setNeutralButton(android.R.string.cancel, doNothing); //cancel button
            AlertDialog dialog = alertDialog.create();
            dialog.show();
            if (diffApps.size() != 0) {
                centerAlignButton(dialog, AlertDialog.BUTTON_POSITIVE);
            }
            if (inconclusiveApps.size() != 0) {
                centerAlignButton(dialog, AlertDialog.BUTTON_NEGATIVE);
            }
            centerAlignButton(dialog, AlertDialog.BUTTON_NEUTRAL);
        }
    };
    //run tomography dialogue
    private final View.OnClickListener runTomoListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            //automatically pops up when tests end if some tests have differentiation
            //asks if user wants to run tomography test
            new AlertDialog.Builder(ReplayActivity.this, AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
                    .setTitle(R.string.localize_diff)
                    .setMessage(R.string.rerun_tomography_descr)
                    .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            Objects.requireNonNull(getSupportActionBar()).setTitle(getString(R.string.tomography_page_title));
                            //rearrange layout to hide rerun button
                            RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)
                                    findViewById(R.id.appsRecyclerView).getLayoutParams();
                            params.addRule(RelativeLayout.ABOVE, R.id.prgBarLayout);
                            findViewById(R.id.rerunButton).setVisibility(View.GONE);
                            findViewById(R.id.localizeDiffButton).setVisibility(View.GONE);
                            //run tomography tests if user clicks yes
                            isTomography = true;
                            adapter.setTomography(true);
                            selectedApps = new ArrayList<>(diffApps);
                            for (ApplicationBean app : selectedApps) {
                                app.setTomography(true);
                                app.setArcepNeedsAlerting(false);
                                app.setAlertFCC(false);
                                app.setStatus(getString(R.string.pending));
                            }
                            traceRunner = new TraceRunAsync(ReplayActivity.this);
                            traceRunner.execute("");
                        }
                    })
                    .setNegativeButton(R.string.no, doNothing).create().show();
        }
    };
    /**
     * Force alert dialog to center align buttons.
     *
     * @param dialog the dialog to align
     * @param button the button to align
     */
    private void centerAlignButton(@NonNull AlertDialog dialog, int button) {
        Button b = dialog.getButton(button);
        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) b.getLayoutParams();
        params.gravity = Gravity.CENTER;
        b.setLayoutParams(params);
    }
    /**
     * When tests are finished and there are tests with differentiation or inconclusive tests, show
     * the rerun button, which allow users to rerun these tests. Also, if there is differentiation,
     * show the Localize Differentiation button to allow users to run tomography tests.
     */
    private void displayRerunTomoButtons() {
        //set rerun button to be visible if differentiation or inconclusive apps
        Button rerunButton = findViewById(R.id.rerunButton);
        rerunButton.setVisibility(View.VISIBLE);
        rerunButton.setOnClickListener(rerunListener);
        //TODO: uncomment to allow users to run tomography tests
        /*
        if (!isTomography && diffApps.size() > 0) { //show tomography button if necessary
            Button runTomoButton = findViewById(R.id.localizeDiffButton);
            runTomoButton.setVisibility(View.VISIBLE);
            runTomoButton.setOnClickListener(runTomoListener);
        }*/
        //rearrange layout so progress bar disappears
        RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams)
                findViewById(R.id.appsRecyclerView).getLayoutParams();
        params.addRule(RelativeLayout.ABOVE, R.id.actionBtnsLayout);
        findViewById(R.id.prgBar).setVisibility(View.GONE);
    }
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        //entry point coming from SelectionFragment
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_replay);
        Toolbar mToolbar = findViewById(R.id.replay_bar);
        setSupportActionBar(mToolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(getString(R.string.replay_page_title));
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        // keep the screen on
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        // First check to see of Internet access is available
        // TODO integrate this with system events like WIFI_STATUS changes same for LTE,
        //  based on events received show appropriate messages to the user
        if (isNetworkUnavailable()) {
            displayNoNetworkDialogue();
        }
        //get stuff from SelectionFragment
        Bundle bundle = getIntent().getExtras();
        assert bundle != null;
        runPortTests = bundle.getBoolean("runPortTests");
        carrier = bundle.getString("carrier");
        selectedApps = getIntent().getParcelableArrayListExtra("selectedApps");
        assert selectedApps != null;
        for (ApplicationBean app : selectedApps) {
            app.setStatus(getString(R.string.pending));
        }
        //set the view for each selected app
        adapter = new ImageReplayRecyclerViewAdapter(selectedApps, this, runPortTests);
        RecyclerView appsRecyclerView = findViewById(R.id.appsRecyclerView);
        RecyclerView.LayoutManager appsRecyclerViewLayoutManager = new LinearLayoutManager(this);
        appsRecyclerView.setLayoutManager(appsRecyclerViewLayoutManager);
        appsRecyclerView.setAdapter(adapter);
        prgBar = findViewById(R.id.prgBar);
        context = getApplicationContext();
        // This is the core of the Application
        if (!isNetworkUnavailable()) {
            traceRunner = new TraceRunAsync(this);
            traceRunner.execute("");
        }
    }
    @Override
    protected void onDestroy() {
        //does this before going back to SelectionFragment
        super.onDestroy();
        if (replayOngoing) {
            if (traceRunner != null) {
                traceRunner.cancel(true);
            }
            Toast.makeText(ReplayActivity.this, getText(R.string.replay_aborted),
                    Toast.LENGTH_LONG).show();
        }
    }
    @Override
    public void onResume() {
        super.onResume();
        adapter.notifyDataSetChanged();
    }
    /**
     * This Method checks the network Availability. For this NetworkInfo class is used and this
     * should also provide type of connectivity i.e. Wi-Fi, Cellular ..
     *
     * @return true if network is available, false otherwise
     */
    public boolean isNetworkUnavailable() {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        // Check if connectivityManager is not null
        if (connectivityManager != null) {
            // Get the active network
            Network activeNetwork = connectivityManager.getActiveNetwork();
            // If there is no active network, the network is unavailable
            if (activeNetwork == null) {
                return true;
            }
            // Get network capabilities and check for connectivity
            NetworkCapabilities networkCapabilities = connectivityManager.getNetworkCapabilities(activeNetwork);
            // Check if the network is connected to Wi-Fi or mobile data
            if (networkCapabilities != null) {
                // Return true if the network is connected to the internet (either Wi-Fi or mobile data)
                return !networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
            }
        }
        // If the connectivityManager is null, consider the network unavailable
        return true;
    }
    /**
     * Display this popup message if there is no network
     */
    public void displayNoNetworkDialogue() {
        ReplayActivity.this.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                new AlertDialog.Builder(ReplayActivity.this,
                        AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
                        .setTitle(getString(R.string.network_error))
                        .setMessage(getString(R.string.text_network_error))
                        .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                replayStop();
                            }
                        }).show();
            }
        });
    }
    /**
     * User wants to leave the replay activity.
     */
    public void replayStop() {
        if (!replayOngoing) {
            ReplayActivity.this.finish(); //calls onDestroy
            ReplayActivity.this.overridePendingTransition(
                    android.R.anim.slide_in_left, android.R.anim.slide_out_right);
        } else {
            new AlertDialog.Builder(ReplayActivity.this, //pop up box
                    AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
                    .setTitle(getString(R.string.interrupt_ongoing_replay_title))
                    .setMessage(getString(R.string.interrupt_ongoing_replay_text))
                    .setPositiveButton(getString(android.R.string.yes), //yes button
                            new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    //calls onDestroy
                                    ReplayActivity.this.finish(); //go back to SelectionFragment
                                    ReplayActivity.this.overridePendingTransition(
                                            android.R.anim.slide_in_left,
                                            android.R.anim.slide_out_right);
                                }
                            })
                    .setNegativeButton(getString(android.R.string.no), doNothing).show(); //no button
        }
    }
    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            replayStop();
            if (!replayOngoing) {
                return super.onKeyDown(keyCode, event);
            } else {
                return true;
            }
        }
        return super.onKeyDown(keyCode, event);
    }
    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        //the left arrow at the top of the screen - user wants to go back
        if (item.getItemId() == android.R.id.home) {
            replayStop();
        }
        return true;
    }

    // Methods for managing replayOngoing state
    void setReplayOngoing(boolean replayOngoing) {
        this.replayOngoing = replayOngoing;
    }

    boolean isReplayOngoing() {
        return replayOngoing;
    }

    // Add app to different lists based on result
    void addToDiffApps(ApplicationBean app) {
        diffApps.add(app);
    }

    void addToInconclusiveApps(ApplicationBean app) {
        inconclusiveApps.add(app);
    }

    // Access to UI elements and app data
    ProgressBar getProgressBar() {
        return prgBar;
    }

    Context getActivityContext() {
        return this;
    }

    ImageReplayRecyclerViewAdapter getAdapter() {
        return adapter;
    }

    ArrayList<ApplicationBean> getSelectedApps() {
        return selectedApps;
    }

    boolean isTomography() {
        return isTomography;
    }

    String getCarrier() {
        return carrier;
    }

    boolean getRunPortTests() {
        return runPortTests;
    }

    ArrayList<ApplicationBean> getDiffApps() {
        return diffApps;
    }

    ArrayList<ApplicationBean> getInconclusiveApps() {
        return inconclusiveApps;
    }

    void displayRerunButtons() {
        displayRerunTomoButtons();
    }
}

File: wehe/src/main/java/mobi/meddle/wehe/ui/replay/ReplayActivity.kt
package mobi.meddle.wehe.ui.replay

import android.content.DialogInterface
import android.os.Bundle
import android.view.Gravity
import android.view.KeyEvent
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.RelativeLayout
import android.widget.Toast
import android.app.AlertDialog
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dagger.hilt.android.AndroidEntryPoint
import mobi.meddle.wehe.R
import mobi.meddle.wehe.adapter.ImageReplayRecyclerViewAdapter
import mobi.meddle.wehe.data.model.ApplicationBean

@AndroidEntryPoint
class ReplayActivity : AppCompatActivity() {

    private val viewModel: ReplayViewModel by viewModels()
    private lateinit var progressBar: ProgressBar
    private lateinit var adapter: ImageReplayRecyclerViewAdapter
    private lateinit var headerLayout: LinearLayout
    private lateinit var headerImage: ImageView
    private lateinit var headerText: TextView
    private val doNothing = DialogInterface.OnClickListener { _, _ -> }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_replay)

        // Setup toolbar
        val mToolbar = findViewById<Toolbar>(R.id.replay_bar)
        setSupportActionBar(mToolbar)
        supportActionBar?.apply {
            title = getString(R.string.replay_page_title)
            setHomeButtonEnabled(true)
            setDisplayHomeAsUpEnabled(true)
        }

        // Keep the screen on
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        // Get extras from intent
        val bundle = intent.extras
        if (bundle != null) {
            val runPortTests = bundle.getBoolean("runPortTests")
            val carrier = bundle.getString("carrier")
            val selectedApps = intent.getParcelableArrayListExtra<ApplicationBean>("selectedApps")

            if (selectedApps != null) {
                // Initialize the ViewModel with data
                viewModel.initializeData(runPortTests, carrier, selectedApps, applicationContext)

                // Setup RecyclerView
                adapter = ImageReplayRecyclerViewAdapter(selectedApps, this, runPortTests)
                val appsRecyclerView = findViewById<RecyclerView>(R.id.appsRecyclerView)
                val layoutManager = LinearLayoutManager(this)
                appsRecyclerView.layoutManager = layoutManager
                appsRecyclerView.adapter = adapter

                // Setup progress bar
                progressBar = findViewById(R.id.prgBar)

                // Check network before starting tests
                if (viewModel.isNetworkUnavailable(this)) {
                    viewModel.showNoNetworkDialog()
                } else {
                    viewModel.execute()
                }
            }
        }

        headerLayout = findViewById(R.id.headerLayout)
        headerImage = findViewById(R.id.headerImage)
        headerText = findViewById(R.id.headerText)

        // Observe LiveData from ViewModel
        setupObservers()
    }

    private fun setupObservers() {
        // Observe isReplayOngoing state
        viewModel.isReplayOngoing.observe(this) { isOngoing ->
            // Update UI based on replay state if needed
        }

        // Observe status updates
        viewModel.statusUpdateEvent.observe(this) { (appName, status) ->
            // Update status of app in adapter
            adapter.notifyDataSetChanged()
        }

        viewModel.currentTestingApp.observe(this) { appInfo ->
            if (appInfo != null) {

                val resourceId = resources.getIdentifier(appInfo.image, "drawable", packageName)

                // Update image based on current app
                if (appInfo.image != null) {
                    headerImage.setImageResource(resourceId)
                    headerLayout.visibility = View.VISIBLE
                }
            }
        }

        viewModel.iteration.observe(this) { iter ->
            if (iter != null) {
                headerText.text = getString(R.string.replay_header_text, iter.toString())
                headerLayout.visibility = View.VISIBLE
            }
        }

        // Observe progress updates
        viewModel.progressUpdateEvent.observe(this) { progress ->
            if (progressBar.visibility == View.GONE || progressBar.visibility == View.INVISIBLE) {
                progressBar.visibility = View.VISIBLE
            }
            progressBar.progress = progress
        }

        // Observe progress completion events
        viewModel.progressCompleteEvent.observe(this) { iteration ->
            if (iteration == 1) {
                progressBar.progress = 50
            } else {
                progressBar.progress = 100
                progressBar.visibility = View.GONE
            }
        }

        // Observe toast messages
        viewModel.toastEvent.observe(this) { message ->
            Toast.makeText(this, message, Toast.LENGTH_LONG).show()
        }

        // Observe dialog events
        viewModel.dialogEvent.observe(this) { (title, message, exitReplays) ->
            AlertDialog.Builder(this, AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(getString(android.R.string.ok)) { _, _ ->
                    if (exitReplays) {
                        replayStop()
                    } else {
                        // All tests just finished, check if we need to display rerun buttons
                        if (viewModel.diffApps.size != 0 || viewModel.inconclusiveApps.size != 0) {
                            viewModel.showRerunTomoButtonsEvent.value?.let {
                                if (it) displayRerunTomoButtons()
                            }
                        }
                    }
                }.show()

            if (exitReplays) {
                supportActionBar?.setTitle(R.string.simple_error)
            } else {
                supportActionBar?.setTitle(R.string.test_results)
            }
        }

        // Observe rerun tomography buttons event
        viewModel.showRerunTomoButtonsEvent.observe(this) { show ->
            if (show) displayRerunTomoButtons()
        }

        // update apps list in adapter
        viewModel.appsList.observe(this) { apps ->
            adapter.updateApps(ArrayList(apps))
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (viewModel.isReplayOngoing.value == true) {
            viewModel.cancel()
            Toast.makeText(
                this, getText(R.string.replay_aborted),
                Toast.LENGTH_LONG
            ).show()
        }
    }

    override fun onResume() {
        super.onResume()
        adapter.notifyDataSetChanged()
    }

    /**
     * User wants to leave the replay activity.
     */
    private fun replayStop() {
        if (viewModel.isReplayOngoing.value != true) {
            finish()
            overridePendingTransition(
                android.R.anim.slide_in_left, android.R.anim.slide_out_right
            )
        } else {
            AlertDialog.Builder(this, AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
                .setTitle(getString(R.string.interrupt_ongoing_replay_title))
                .setMessage(getString(R.string.interrupt_ongoing_replay_text))
                .setPositiveButton(getString(android.R.string.yes)) { _, _ ->
                    finish()
                    overridePendingTransition(
                        android.R.anim.slide_in_left,
                        android.R.anim.slide_out_right
                    )
                }
                .setNegativeButton(getString(android.R.string.no), doNothing)
                .show()
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            replayStop()
            return if (viewModel.isReplayOngoing.value != true) {
                super.onKeyDown(keyCode, event)
            } else {
                true
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            replayStop()
        }
        return true
    }

    /**
     * Display buttons for rerunning tests or conducting tomography tests
     */
    private fun displayRerunTomoButtons() {
        // Set rerun button to be visible if there are differentiation or inconclusive apps
        val rerunButton = findViewById<Button>(R.id.rerunButton)
        rerunButton.visibility = View.VISIBLE
        rerunButton.setOnClickListener { showRerunDialog() }

//        // Show tomography button if there are apps with differentiation and not in tomography mode
//        if (!viewModel.isTomography && viewModel.diffApps.size > 0) {
//            val runTomoButton = findViewById<Button>(R.id.localizeDiffButton)
//            runTomoButton.visibility = View.VISIBLE
//            runTomoButton.setOnClickListener { showTomographyDialog() }
//        }

        // Rearrange layout so progress bar disappears
        val params = findViewById<View>(R.id.appsRecyclerView).layoutParams as RelativeLayout.LayoutParams
        params.addRule(RelativeLayout.ABOVE, R.id.actionBtnsLayout)
        findViewById<View>(R.id.prgBar).visibility = View.GONE
    }

    /**
     * Show dialog to rerun tests
     */
    private fun showRerunDialog() {
        val alertDialog = AlertDialog.Builder(this, AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
            .setTitle(R.string.rerun_test_title)
            .setMessage(R.string.rerun_test_descr)

        // Rerun tests with differentiation
        if (viewModel.diffApps.size != 0) {
            alertDialog.setPositiveButton(R.string.rerun_diff_opt) { _, _ ->
                prepareForRerun(true)
            }
        }

        // Rerun only the inconclusive tests
        if (viewModel.inconclusiveApps.size != 0) {
            alertDialog.setNegativeButton(R.string.rerun_incon_opt) { _, _ ->
                prepareForRerun(false)
            }
        }

        alertDialog.setNeutralButton(android.R.string.cancel, doNothing)
        val dialog = alertDialog.create()
        dialog.show()

        // Center align buttons
        if (viewModel.diffApps.size != 0) {
            centerAlignButton(dialog, AlertDialog.BUTTON_POSITIVE)
        }
        if (viewModel.inconclusiveApps.size != 0) {
            centerAlignButton(dialog, AlertDialog.BUTTON_NEGATIVE)
        }
        centerAlignButton(dialog, AlertDialog.BUTTON_NEUTRAL)
    }

    /**
     * Prepare UI and ViewModel for rerun tests
     */
    private fun prepareForRerun(isRunningDifferentiation: Boolean) {
        // Change page title
        supportActionBar?.title = getString(R.string.replay_page_title)

        // Rearrange layout to hide rerun button
        val params = findViewById<View>(R.id.appsRecyclerView).layoutParams as RelativeLayout.LayoutParams
        params.addRule(RelativeLayout.BELOW, R.id.prgBarLayout)
        findViewById<View>(R.id.rerunButton).visibility = View.GONE
        findViewById<View>(R.id.localizeDiffButton).visibility = View.GONE

        // Update adapter and prepare viewModel
        adapter.setTomography(false)
        val newApps = viewModel.prepareRerunTests(isRunningDifferentiation)
        adapter.updateApps(newApps)
        viewModel.execute()
    }

//    /**
//     * Show dialog to run tomography tests
//     */
//    private fun showTomographyDialog() {
//        AlertDialog.Builder(this, AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
//            .setTitle(R.string.localize_diff)
//            .setMessage(R.string.rerun_tomography_descr)
//            .setPositiveButton(R.string.yes) { _, _ ->
//                prepareTomographyTests()
//            }
//            .setNegativeButton(R.string.no, doNothing)
//            .create()
//            .show()
//    }

//    /**
//     * Prepare UI and ViewModel for tomography tests
//     */
//    private fun prepareTomographyTests() {
//        // Change page title
//        supportActionBar?.title = getString(R.string.tomography_page_title)
//
//        // Rearrange layout to hide rerun button
//        val params = findViewById<View>(R.id.appsRecyclerView).layoutParams as RelativeLayout.LayoutParams
//        params.addRule(RelativeLayout.ABOVE, R.id.prgBarLayout)
//        findViewById<View>(R.id.rerunButton).visibility = View.GONE
//        findViewById<View>(R.id.localizeDiffButton).visibility = View.GONE
//
//        // Update adapter for tomography tests
//        adapter.setTomography(true)
//
//        // Prepare and execute tomography tests
//        viewModel.prepareTomographyTests()
//    }

    /**
     * Force alert dialog to center align buttons.
     */
    private fun centerAlignButton(dialog: AlertDialog, button: Int) {
        val b = dialog.getButton(button)
        val params = b.layoutParams as LinearLayout.LayoutParams
        params.gravity = Gravity.CENTER
        b.layoutParams = params
    }

    companion object {
        const val STATUS: String = "ReplayActPrefsFile"
    }
}